# Bay Area, CA - Airbnb Data


## Context

Since its inception in 2008, Airbnb has disrupted the hospitality industry by allowing almost anyone to rent out a spare room and host travelers looking for an overnight stay. 

While Airbnb has publicly available data from many locations, **Santa Clara County is of particular interest as it is one of the major counties in the Bay Area and home to many prominent tech companies and startups in the Silicon Valley.**
As such, home prices and cost of living are exceptionally high in this area, as one will see in the data.


## Content

All data is publicly available under the Creative Commons "Public Domain Dedication" license and has been updated as of June 12th, 2020.Provided are metrics that are publicly visible for each listing (e.g. name, description, price, reviews, etc.), but the dataset contains some Airbnb's internal metrics as well (review score accuracy, host acceptance rate, etc.).

## Acknowledgements

Acknowledgments go to Airbnb for their publicly released datasets that are available at this website.




## Inspiration & ideas:

1. Create a price-suggestion model for new Airbnb hosts who might not know the value of their listing.
    - Text mining with statified house price data
2. Can we predict the rating of an Airbnb listing utilizing **NLP** of the description columns?How have Airbnb
3. prices changed over time? Are prices seasonal?
   - Time series analysis?
4. Which areas in the Santa Clara County are most "popular"?
   - Region analysis
5. Which features of an Airbnb listing are important to add to its perceived value?
   - Feature analysis,PCA
   - Correlation analysis
   

## Import  libraries and data

In [77]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

In [78]:
data = pd.read_csv('./data/Airbnb_Listings.csv')
data.head()

id                         listing_url       scrape_id last_scraped  \
0   4952   https://www.airbnb.com/rooms/4952  20200530151957   2020-05-30   
1  11464  https://www.airbnb.com/rooms/11464  20200530151957   2020-05-31   
2  17884  https://www.airbnb.com/rooms/17884  20200530151957   2020-05-31   
3  21373  https://www.airbnb.com/rooms/21373  20200530151957   2020-05-30   
4  37512  https://www.airbnb.com/rooms/37512  20200530151957   2020-05-31   

                                  name  \
0    Butterfly Inn - Graceful Living !   
1    Deluxe Private Studio-custom int.   
2                 Silicon Valley Suite   
3  Bonsai Garden Inn in Professorville   
4    Private  room - Parking 3 carport   

                                             summary  \
0  Lovely garden setting in a serene and art-fill...   
1  Custom built Studio with exquisite design. Rea...   
2  A guest suite for one or two, in a house in a ...   
3  Room  in gracious home with beautiful garden. ...   
4  We live in a safe community close to public tr...   

                                               space  \
0  Very comfortable Queen bed and small desk in b...   
1  Description A favorite for international corpo...   
2  This is a private suite at the rear of a house...   
3  Bright, garden-facing room in beautiful home. ...   
4  I have a really nice room in a quiet neighborh...   

                                         description experiences_offered  \
0  Lovely garden setting in a serene and art-fill...                none   
1  Custom built Studio with exquisite design. Rea...                none   
2  A guest suite for one or two, in a house in a ...                none   
3  Room  in gracious home with beautiful garden. ...                none   
4  We live in a safe community close to public tr...                none   

                               neighborhood_overview  ... instant_bookable  \
0  Located in Professorville, Palo Alto, one of t...  ...                f   
1  10 min. to Lucky's and Starbucks at El Camino ...  ...                f   
2  This is a very quiet family neighborhood, but ...  ...                f   
3  This room is in an ultra convenient location i...  ...                f   
4  Our community is a safe environment and at nig...  ...                f   

  is_business_travel_ready          cancellation_policy  \
0                        f                     moderate   
1                        f                     moderate   
2                        f  strict_14_with_grace_period   
3                        f                     moderate   
4                        f                     moderate   

  require_guest_profile_picture require_guest_phone_verification  \
0                             f                                f   
1                             f                                t   
2                             f                                f   
3                             f                                f   
4                             t                                t   

   calculated_host_listings_count  \
0                               5   
1                              12   
2                               2   
3                               5   
4                               2   

   calculated_host_listings_count_entire_homes  \
0                                            0   
1                                           12   
2                                            2   
3                                            0   
4                                            0   

  calculated_host_listings_count_private_rooms  \
0                                            5   
1                                            0   
2                                            0   
3                                            5   
4                                            2   

   calculated_host_listings_count_shared_rooms  reviews_per_month  
0                                

In [79]:
data.shape

(7221, 106)

In [80]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7221 entries, 0 to 7220
Columns: 106 entries, id to reviews_per_month
dtypes: float64(23), int64(21), object(62)
memory usage: 5.8+ MB


## Data cleaning

In [81]:
# find missing values and single value columns in data:
del_cols = []

for col in data.columns:
    if data[col].isna().all():
        del_cols.append(col)
    elif data[col].nunique() == 1:
        del_cols.append(col)

In [82]:
del_cols

['scrape_id',
 'experiences_offered',
 'thumbnail_url',
 'medium_url',
 'xl_picture_url',
 'neighbourhood_group_cleansed',
 'country_code',
 'country',
 'has_availability',
 'requires_license',
 'license',
 'is_business_travel_ready']

In [83]:
data = data.drop(del_cols, axis=1)

In [84]:
data['id'].isna().sum()

0

In [85]:
# delete url column:
urls = [
    'listing_url', 'picture_url', 'host_url', 'host_thumbnail_url',
    'host_picture_url'
]


In [86]:
data = data.drop(urls, axis=1)

In [87]:
for col in data.columns:
    list_unique = (col, len(data[col].unique()))
    print(list_unique)

('id', 7221)
('last_scraped', 2)
('name', 7078)
('summary', 6052)
('space', 4403)
('description', 6440)
('neighborhood_overview', 3509)
('notes', 2747)
('transit', 3200)
('access', 3378)
('interaction', 3166)
('house_rules', 3718)
('host_id', 3533)
('host_name', 2103)
('host_since', 2085)
('host_location', 242)
('host_about', 2030)
('host_response_time', 5)
('host_response_rate', 44)
('host_acceptance_rate', 78)
('host_is_superhost', 3)
('host_neighbourhood', 145)
('host_listings_count', 62)
('host_total_listings_count', 62)
('host_verifications', 240)
('host_has_profile_pic', 3)
('host_identity_verified', 3)
('street', 37)
('neighbourhood', 37)
('neighbourhood_cleansed', 16)
('city', 35)
('state', 5)
('zipcode', 83)
('market', 5)
('smart_location', 37)
('latitude', 5882)
('longitude', 6357)
('is_location_exact', 2)
('property_type', 28)
('room_type', 4)
('accommodates', 16)
('bathrooms', 16)
('bedrooms', 11)
('beds', 21)
('bed_type', 5)
('amenities', 6396)
('square_feet', 25)
('price'

In [88]:
data['property_type'].value_counts()

House                 3715
Apartment             1152
Serviced apartment     504
Townhouse              429
Guest suite            363
Guesthouse             318
Condominium            303
Villa                  129
Bungalow               115
Loft                    41
Cottage                 29
Camper/RV               27
Boutique hotel          23
Tiny house              16
Other                   13
Bed and breakfast       12
Cabin                    7
Farm stay                6
Tent                     5
Treehouse                3
Yurt                     3
Barn                     2
Train                    1
Lighthouse               1
Campsite                 1
Earth house              1
Aparthotel               1
Chalet                   1
Name: property_type, dtype: int64

In [89]:
data['state'].unique() # to drop

array(['CA', 'Ca', 'ca', nan, 'California '], dtype=object)

In [90]:
data_null = data.isnull().sum()

In [91]:
data_null.sort_values(ascending=False)

square_feet               7192
weekly_price              6647
monthly_price             6629
notes                     3529
access                    3117
                          ... 
maximum_minimum_nights       0
minimum_maximum_nights       0
maximum_maximum_nights       0
minimum_nights_avg_ntm       0
bed_type                     0
Length: 89, dtype: int64

In [92]:
data = data.drop(['square_feet','zipcode','state'], axis=1)

In [93]:
data.columns

Index(['id', 'last_scraped', 'name', 'summary', 'space', 'description',
       'neighborhood_overview', 'notes', 'transit', 'access', 'interaction',
       'house_rules', 'host_id', 'host_name', 'host_since', 'host_location',
       'host_about', 'host_response_time', 'host_response_rate',
       'host_acceptance_rate', 'host_is_superhost', 'host_neighbourhood',
       'host_listings_count', 'host_total_listings_count',
       'host_verifications', 'host_has_profile_pic', 'host_identity_verified',
       'street', 'neighbourhood', 'neighbourhood_cleansed', 'city', 'market',
       'smart_location', 'latitude', 'longitude', 'is_location_exact',
       'property_type', 'room_type', 'accommodates', 'bathrooms', 'bedrooms',
       'beds', 'bed_type', 'amenities', 'price', 'weekly_price',
       'monthly_price', 'security_deposit', 'cleaning_fee', 'guests_included',
       'extra_people', 'minimum_nights', 'maximum_nights',
       'minimum_minimum_nights', 'maximum_minimum_nights',
       '

### Transfer the unit of currency to float

In [94]:
price_list = [
    'weekly_price',
    'monthly_price',
    'price',
    'security_deposit',
    'cleaning_fee',
    'extra_people',
]

In [95]:
for p in price_list: 
    print(p,data[p].isna().sum())

weekly_price 6647
monthly_price 6629
price 0
security_deposit 1457
cleaning_fee 800
extra_people 0


In [96]:
for p in price_list:
    data[p].fillna('$0.00', inplace=True)
    data[p] = data[p].str.replace('$', '').str.replace(',', '').astype(float)

/var/folders/dv/qnpc2tm905n3c2nrmq03c65w0000gn/T/ipykernel_60738/1242419390.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data[p] = data[p].str.replace('$', '').str.replace(',', '').astype(float)


In [97]:
data[price_list].isna().sum()

weekly_price        0
monthly_price       0
price               0
security_deposit    0
cleaning_fee        0
extra_people        0
dtype: int64

In [98]:
data.describe(include='all')

id last_scraped  \
count   7.221000e+03         7221   
unique           NaN            2   
top              NaN   2020-05-31   
freq             NaN         4782   
mean    2.623922e+07          NaN   
std     1.236420e+07          NaN   
min     4.952000e+03          NaN   
25%     1.624242e+07          NaN   
50%     2.818432e+07          NaN   
75%     3.749948e+07          NaN   
max     4.359134e+07          NaN   

                                                    name  \
count                                               7220   
unique                                              7077   
top     TRIBE ❤️ Deep Cleaned  2BR @ Great Mall Milpitas   
freq                                                  21   
mean                                                 NaN   
std                                                  NaN   
min                                                  NaN   
25%                                                  NaN   
50%                                                  NaN   
75%                                                  NaN   
max                                                  NaN   

                                                  summary  \
count                                                6922   
unique                                               6051   
top     The HackerHome Network New to the Valley? Tran...   
freq                                                   44   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                                    space  \
count                                                5514   
unique                                               4402   
top     At Zeus, we focus exclusively on 30 day+ stays...   
freq                                                  411   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                              description  \
count                                                6983   
unique                                               6439   
top     Dorm-style house in Silicon Valley. No-host, 5...   
freq                                                   20   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                    neighborhood_overview  \
count                                                4544   
unique                                               3508   
top     *Area Highlights* -Located in the Heart of the...   
freq                                                   31   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                        

## Data Exploartion

### Check the numerical and nominal data seperately

In [99]:
numeric_features = data.select_dtypes(include=['int64', 'float64'])
nominal_features = data.select_dtypes(include=['object'])

In [100]:
numeric_features.head()

id  host_id  host_listings_count  host_total_listings_count  latitude  \
0   4952     7054                  5.0                        5.0  37.43972   
1  11464    42458                 13.0                       13.0  37.34342   
2  17884    69023                  2.0                        2.0  37.38612   
3  21373     7054                  5.0                        5.0  37.44001   
4  37512   161741                  2.0                        2.0  37.30598   

   longitude  accommodates  bathrooms  bedrooms  beds  ...  \
0 -122.15672             2        2.0       1.0   1.0  ...   
1 -121.98829             1        1.0       NaN   1.0  ...   
2 -122.07357             2        1.0       1.0   1.0  ...   
3 -122.15667             1        2.0       1.0   1.0  ...   
4 -121.84036             2        1.0       1.0   1.0  ...   

   review_scores_cleanliness  review_scores_checkin  \
0                       10.0                   10.0   
1                       10.0                   10.0   
2                       10.0                   10.0   
3                       10.0                   10.0   
4                       10.0                   10.0   

   review_scores_communication  review_scores_location  review_scores_value  \
0                         10.0                    10.0                 10.0   
1                          9.0                     9.0                  9.0   
2                         10.0                    10.0                 10.0   
3                         10.0                    10.0                 10.0   
4                         10.0                     9.0                  9.0   

   calculated_host_listings_count  \
0                               5   
1                              12   
2                               2   
3                               5   
4                               2   

   calculated_host_listings_count_entire_homes  \
0                                            0   
1                                           12   
2                                            2   
3                                            0   
4                                            0   

   calculated_host_listings_count_private_rooms  \
0                                             5   
1                                             0   
2                                             0   
3                                             5   
4                                             2   

   calculated_host_listings_count_shared_rooms  reviews_per_month  
0                                            0               0.57  
1                                            0               0.08  
2                                            0               0.11  
3                                            0               2.17  
4                                            0               1.53  

[5 rows x 43 columns]

In [101]:
numeric_features.columns

Index(['id', 'host_id', 'host_listings_count', 'host_total_listings_count',
       'latitude', 'longitude', 'accommodates', 'bathrooms', 'bedrooms',
       'beds', 'price', 'weekly_price', 'monthly_price', 'security_deposit',
       'cleaning_fee', 'guests_included', 'extra_people', 'minimum_nights',
       'maximum_nights', 'minimum_minimum_nights', 'maximum_minimum_nights',
       'minimum_maximum_nights', 'maximum_maximum_nights',
       'minimum_nights_avg_ntm', 'maximum_nights_avg_ntm', 'availability_30',
       'availability_60', 'availability_90', 'availability_365',
       'number_of_reviews', 'number_of_reviews_ltm', 'review_scores_rating',
       'review_scores_accuracy', 'review_scores_cleanliness',
       'review_scores_checkin', 'review_scores_communication',
       'review_scores_location', 'review_scores_value',
       'calculated_host_listings_count',
       'calculated_host_listings_count_entire_homes',
       'calculated_host_listings_count_private_rooms',
       'cal

In [204]:
(data['maximum_maximum_nights'] ==  data['minimum_maximum_nights']).value_counts()

True     7178
False      43
dtype: int64

In [103]:
(data['host_listings_count'] == data['host_total_listings_count']).value_counts()

True     7220
False       1
dtype: int64

In [104]:
numeric_features.isna().sum()

id                                                 0
host_id                                            0
host_listings_count                                1
host_total_listings_count                          1
latitude                                           0
longitude                                          0
accommodates                                       0
bathrooms                                          1
bedrooms                                           6
beds                                              68
price                                              0
weekly_price                                       0
monthly_price                                      0
security_deposit                                   0
cleaning_fee                                       0
guests_included                                    0
extra_people                                       0
minimum_nights                                     0
maximum_nights                                

In [105]:
nominal_features.head()

last_scraped                                 name  \
0   2020-05-30    Butterfly Inn - Graceful Living !   
1   2020-05-31    Deluxe Private Studio-custom int.   
2   2020-05-31                 Silicon Valley Suite   
3   2020-05-30  Bonsai Garden Inn in Professorville   
4   2020-05-31    Private  room - Parking 3 carport   

                                             summary  \
0  Lovely garden setting in a serene and art-fill...   
1  Custom built Studio with exquisite design. Rea...   
2  A guest suite for one or two, in a house in a ...   
3  Room  in gracious home with beautiful garden. ...   
4  We live in a safe community close to public tr...   

                                               space  \
0  Very comfortable Queen bed and small desk in b...   
1  Description A favorite for international corpo...   
2  This is a private suite at the rear of a house...   
3  Bright, garden-facing room in beautiful home. ...   
4  I have a really nice room in a quiet neighborh...   

                                         description  \
0  Lovely garden setting in a serene and art-fill...   
1  Custom built Studio with exquisite design. Rea...   
2  A guest suite for one or two, in a house in a ...   
3  Room  in gracious home with beautiful garden. ...   
4  We live in a safe community close to public tr...   

                               neighborhood_overview  \
0  Located in Professorville, Palo Alto, one of t...   
1  10 min. to Lucky's and Starbucks at El Camino ...   
2  This is a very quiet family neighborhood, but ...   
3  This room is in an ultra convenient location i...   
4  Our community is a safe environment and at nig...   

                                               notes  \
0  Housekeeping every Monday, leave your door ope...   
1  Pet Policy: - Well Behaved pet up to 25 lb. of...   
2  I cannot accommodate cats, sorry. I can usuall...   
3  The family room has a flat panel tv and desk f...   
4  Please remember that you are in my home and re...   

                                             transit  \
0  Walking distance to Stanford University (30 mi...   
1  Public transportation at Homestead and Pomeroy...   
2  The CalTrain station is in walking distance, a...   
3  The Stanford Shopping Center is 20 minutes wal...   
4  We have a bus stop right outside our community...   

                                              access  \
0  Kitchen, laundry, family/TV room, garden, free...   
1          Complimentary Wifi-internet + Basic Cable   
2  Private outdoor patio. Shared washer/dryer on ...   
3  Kitchen, Laundry, Garden, Family Room with TV,...   
4  The kitchen and all that it offers.  The laund...   

                                         interaction  ...  \
0  Welcome and Orientation. Shared kitchen, garde...  ...   
1  Available any time for support and problem res...  ...   
2  I am usually in town.  Guests have a private e...  ...   
3  We will see you everyday as we cook meals in t...  ...   
4  We follow our guests lead when it comes to soc...  ...   

                                           amenities calendar_updated  \
0  {TV,"Cable TV",Internet,Wifi,Kitchen,Heating,W...     2 months ago   
1  {TV,"Cable TV",Internet,Wifi,"Air conditioning...     5 months ago   
2  {TV,"Cable TV",Internet,Wifi,"Air conditioning...      4 weeks ago   
3  {TV,"Cable TV",Internet,Wifi,Kitchen,"Paid par...      3 weeks ago   
4  {Internet,Wifi,"Air conditioning",Kitchen,"Fre...    23 months ago   

  calendar_last_scraped first_review last_review       jurisdiction_names  \
0            2020-05-30   2009-08-02  2019-01-12            {"PALO ALTO"}   
1            2020-05-31   2014-06-15  2019-10-11          {"SANTA CLARA"}   
2            2020-05-31   2018-11-20  2019-04-27  {"Mountain View"," CA"}   
3            2020-05-30   2010-05-28  2020-03-14            {"PALO ALTO"}   
4            2020-05-31   2010-08-01  2017-03-14             {"SAN JOSE"}   

  instant_bookable          cancellation_po

In [106]:
nominal_features.columns

Index(['last_scraped', 'name', 'summary', 'space', 'description',
       'neighborhood_overview', 'notes', 'transit', 'access', 'interaction',
       'house_rules', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_neighbourhood', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'street',
       'neighbourhood', 'neighbourhood_cleansed', 'city', 'market',
       'smart_location', 'is_location_exact', 'property_type', 'room_type',
       'bed_type', 'amenities', 'calendar_updated', 'calendar_last_scraped',
       'first_review', 'last_review', 'jurisdiction_names', 'instant_bookable',
       'cancellation_policy', 'require_guest_profile_picture',
       'require_guest_phone_verification'],
      dtype='object')

In [107]:
(nominal_features['last_scraped'] == nominal_features['calendar_last_scraped']).value_counts()

True    7221
dtype: int64

In [108]:
nominal_features.isna().sum()

last_scraped                           0
name                                   1
summary                              299
space                               1707
description                          238
neighborhood_overview               2677
notes                               3529
transit                             3085
access                              3117
interaction                         2968
house_rules                         2264
host_name                              1
host_since                             1
host_location                         23
host_about                          2334
host_response_time                  1807
host_response_rate                  1807
host_acceptance_rate                 714
host_is_superhost                      1
host_neighbourhood                  1181
host_verifications                     0
host_has_profile_pic                   1
host_identity_verified                 1
street                                 0
neighbourhood   

In [109]:
data['room_type'].unique()

array(['Private room', 'Entire home/apt', 'Shared room', 'Hotel room'],
      dtype=object)

In [110]:
accomo_type = data['accommodates'].unique()
accomo_type.sort()
accomo_type

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16])

In [111]:
data['bed_type'].unique()

array(['Real Bed', 'Futon', 'Pull-out Sofa', 'Airbed', 'Couch'],
      dtype=object)

In [112]:
data['jurisdiction_names'].unique()

array(['{"PALO ALTO"}', '{"SANTA CLARA"}', '{"Mountain View"," CA"}',
       '{"SAN JOSE"}', '{Cupertino," CA"}', '{Sunnyvale," CA"}', nan,
       '{Milpitas," CA"}', '{"Morgan Hill"," CA"}', '{"Los Gatos"," CA"}',
       '{"San Benito County"," CA"}'], dtype=object)

In [113]:
data['jurisdiction_names'].isna().sum()

979

In [114]:
data['cancellation_policy'].unique()

array(['moderate', 'strict_14_with_grace_period', 'flexible',
       'super_strict_60'], dtype=object)

### clean the city column

In [115]:
data['city'].unique()

array(['Palo Alto', 'Santa Clara', 'Mountain View', 'San Jose',
       'Cupertino', 'Sunnyvale', 'Campbell', 'Milpitas', 'Saratoga',
       'Morgan Hill', 'Los Altos', 'Los Gatos', 'Menlo Park',
       'Los Altos Hills', 'Monte Sereno', 'Hollister', 'Gilroy',
       'San Jose ', 'San Martin', 'Stanford', 'Santa Clara County',
       'Sunnyvale ', 'Palo Alto ', 'Milpitas ', 'Los Gatos ', 'san jose',
       'Mountain View ', 'Danville ', nan, 'Campbell ', '洛斯阿尔托斯',
       'santa clara', 'Fremont', '圣何塞', 'Watsonville'], dtype=object)

In [175]:
data['city'] = data['city'].str.replace('圣何塞', 'San Jose')
data['city'] = data['city'].str.replace('san jose', 'San Jose')
data['city'] = data['city'].str.replace('洛斯阿尔托斯', 'Los Altos')
data['city'] = data['city'].str.replace('santa clara', 'Santa Clara')
data['city'] = data['city'].str.replace('Santa Clara County','Santa Clara')
data['city'].str.strip()

0           Palo Alto
1         Santa Clara
2       Mountain View
3           Palo Alto
4            San Jose
            ...      
7216         San Jose
7217      Santa Clara
7218        Sunnyvale
7219         San Jose
7220         San Jose
Name: city, Length: 7221, dtype: object

In [203]:
data['city'].unique()

array(['Palo Alto', 'Santa Clara', 'Mountain View', 'San Jose',
       'Cupertino', 'Sunnyvale', 'Campbell', 'Milpitas', 'Saratoga',
       'Morgan Hill', 'Los Altos', 'Los Gatos', 'Menlo Park',
       'Los Altos Hills', 'Monte Sereno', 'Hollister', 'Gilroy',
       'San Jose ', 'San Martin', 'Stanford', 'Sunnyvale ', 'Palo Alto ',
       'Milpitas ', 'Los Gatos ', 'Mountain View ', 'Danville ', nan,
       'Campbell ', 'Fremont', 'Watsonville'], dtype=object)

In [177]:
# find binary columns:
for col in data.columns:
    if data[col].nunique() == 2:
        print(col)

last_scraped
host_is_superhost
host_has_profile_pic
host_identity_verified
is_location_exact
calendar_last_scraped
instant_bookable
require_guest_profile_picture
require_guest_phone_verification


In [178]:
for col in data.columns:
    if data[col].nunique() == 3:
        print(col)

In [179]:
for col in data.columns:
    if data[col].nunique() == 4:
        print(col)

host_response_time
market
room_type
cancellation_policy


In [180]:
for col in data.columns:
    if data[col].nunique() == 5:
        print(col)

bed_type


In [181]:
for col, n in zip(data.columns, data.nunique()):
    if n <10 and n > 1:
        print(col,n)

last_scraped 2
host_response_time 4
host_is_superhost 2
host_has_profile_pic 2
host_identity_verified 2
market 4
is_location_exact 2
room_type 4
bed_type 5
calendar_last_scraped 2
review_scores_accuracy 8
review_scores_cleanliness 8
review_scores_checkin 8
review_scores_communication 8
review_scores_location 7
review_scores_value 8
instant_bookable 2
cancellation_policy 4
require_guest_profile_picture 2
require_guest_phone_verification 2


## What we got now:

Data cleaning

  - id？ host ID？ which to keep?
  - how to deal with missing price? -- fill na zero?
  - can we delete one of these columns each?
    - (data['maximum_maximum_nights'] ==  data['minimum_maximum_nights']).value_counts()
    - (data['host_listings_count'] == data['host_total_listings_count']).value_counts()
    - (nominal_features['last_scraped'] == nominal_features['calendar_last_scraped']).value_counts()

---
- Neighbors does not match the neighbors in data
  - sub data by region?
- Location :  clean later
  -  ('latitude', 5882)
  -  ('longitude', 6357)

 

In [206]:
data.drop([
    'minimum_maximum_nights', 'host_total_listings_count',
    'calendar_last_scraped'
],
          axis=1,
          inplace=True)

### Amenities hot words analysis

In [182]:
from collections import Counter

results = Counter()

In [183]:
amenities = data['amenities']

In [184]:
amenities.str.strip('{}')\
               .str.replace('"', '')\
               .str.lstrip('\"')\
               .str.rstrip('\"')\
               .str.split(',')\
               .apply(results.update)

0       None
1       None
2       None
3       None
4       None
        ... 
7216    None
7217    None
7218    None
7219    None
7220    None
Name: amenities, Length: 7221, dtype: object

In [185]:
# results

In [186]:
sub_df = pd.DataFrame(results.most_common(50), columns=['amenities', 'count'])

In [187]:
sub_df['ratio'] = round(sub_df['count'] / len(amenities),4)

In [188]:
sub_df.sort_values(by=['count'], ascending=False)

amenities  count   ratio
0                         Wifi   7071  0.9792
1                   Essentials   6904  0.9561
2                      Heating   6824  0.9450
3                  Smoke alarm   6794  0.9409
4                      Hangers   6480  0.8974
5                      Shampoo   6271  0.8684
6        Carbon monoxide alarm   6178  0.8556
7    Laptop-friendly workspace   5998  0.8306
8                   Hair dryer   5997  0.8305
9                      Kitchen   5890  0.8157
10                      Washer   5817  0.8056
11                       Dryer   5785  0.8011
12                        Iron   5711  0.7909
13                   Hot water   5280  0.7312
14    Free parking on premises   5277  0.7308
15                          TV   5203  0.7205
16            Air conditioning   4565  0.6322
17                Refrigerator   4561  0.6316
18                   Microwave   4541  0.6289
19       Dishes and silverware   4139  0.5732
20         Free street parking   4043  0.5599
21           Fire extinguisher   4040  0.5595
22                Coffee maker   3907  0.5411
23                  Bed linens   3710  0.5138
24        Lock on bedroom door   3677  0.5092
25               First aid kit   3668  0.5080
26                       Stove   3455  0.4785
27              Cooking basics   3447  0.4774
28                        Oven   3434  0.4756
29               Self check-in   3257  0.4510
30            Private entrance   3052  0.4227
31  Extra pillows and blankets   3024  0.4188
32     Long term stays allowed   2851  0.3948
33                  Dishwasher   2829  0.3918
34          Garden or backyard   2437  0.3375
35            Patio or balcony   2291  0.3173
36                    Cable TV   2228  0.3085
37     Luggage dropoff allowed   1839  0.2547
38                    Internet   1838  0.2545
39           Single level home   1512  0.2094
40                        Pool   1437  0.1990
41         Family/kid friendly   1429  0.1979
42            Indoor fireplace   1345  0.1863
43                      Keypad   1286  0.1781
44         Private living room   1203  0.1666
45                     Lockbox   1197  0.1658
46                     Bathtub   1070  0.1482
47                Pets allowed   1040  0.1440
48                   BBQ grill   1028  0.1424
49                         Gym   1008  0.1396

## Fusion with reviews

In [189]:
reviews = pd.read_csv('./data/reviews.csv')
print(reviews.shape)
reviews.head()

(212513, 2)


listing_id        date
0        4952  2009-08-02
1        4952  2009-09-04
2        4952  2009-10-16
3        4952  2009-12-10
4        4952  2010-06-08

In [190]:
rev_row = reviews['listing_id'].unique()

In [191]:
data_rev = data.loc[data['id'].isin(rev_row)]
data_rev.head()

id last_scraped                                 name  \
0   4952   2020-05-30    Butterfly Inn - Graceful Living !   
1  11464   2020-05-31    Deluxe Private Studio-custom int.   
2  17884   2020-05-31                 Silicon Valley Suite   
3  21373   2020-05-30  Bonsai Garden Inn in Professorville   
4  37512   2020-05-31    Private  room - Parking 3 carport   

                                             summary  \
0  Lovely garden setting in a serene and art-fill...   
1  Custom built Studio with exquisite design. Rea...   
2  A guest suite for one or two, in a house in a ...   
3  Room  in gracious home with beautiful garden. ...   
4  We live in a safe community close to public tr...   

                                               space  \
0  Very comfortable Queen bed and small desk in b...   
1  Description A favorite for international corpo...   
2  This is a private suite at the rear of a house...   
3  Bright, garden-facing room in beautiful home. ...   
4  I have a really nice room in a quiet neighborh...   

                                         description  \
0  Lovely garden setting in a serene and art-fill...   
1  Custom built Studio with exquisite design. Rea...   
2  A guest suite for one or two, in a house in a ...   
3  Room  in gracious home with beautiful garden. ...   
4  We live in a safe community close to public tr...   

                               neighborhood_overview  \
0  Located in Professorville, Palo Alto, one of t...   
1  10 min. to Lucky's and Starbucks at El Camino ...   
2  This is a very quiet family neighborhood, but ...   
3  This room is in an ultra convenient location i...   
4  Our community is a safe environment and at nig...   

                                               notes  \
0  Housekeeping every Monday, leave your door ope...   
1  Pet Policy: - Well Behaved pet up to 25 lb. of...   
2  I cannot accommodate cats, sorry. I can usuall...   
3  The family room has a flat panel tv and desk f...   
4  Please remember that you are in my home and re...   

                                             transit  \
0  Walking distance to Stanford University (30 mi...   
1  Public transportation at Homestead and Pomeroy...   
2  The CalTrain station is in walking distance, a...   
3  The Stanford Shopping Center is 20 minutes wal...   
4  We have a bus stop right outside our community...   

                                              access  ...  \
0  Kitchen, laundry, family/TV room, garden, free...  ...   
1          Complimentary Wifi-internet + Basic Cable  ...   
2  Private outdoor patio. Shared washer/dryer on ...  ...   
3  Kitchen, Laundry, Garden, Family Room with TV,...  ...   
4  The kitchen and all that it offers.  The laund...  ...   

        jurisdiction_names instant_bookable          cancellation_policy  \
0            {"PALO ALTO"}                f                     moderate   
1          {"SANTA CLARA"}                f                     moderate   
2  {"Mountain View"," CA"}                f  strict_14_with_grace_period   
3            {"PALO ALTO"}                f                     moderate   
4             {"SAN JOSE"}                f                     moderate   

  require_guest_profile_picture require_guest_phone_verification  \
0                             f                                f   
1                             f                                t   
2                             f                                f   
3                             f                                f   
4                             t                                t   

  calculated_host_listings_count calculated_host_listings_count_entire_homes  \
0                              5                                           0   
1                             12                                          12   
2                              2                                           2   
3                              5                          

In [192]:
data_rev.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5734 entries, 0 to 7214
Data columns (total 86 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            5734 non-null   int64  
 1   last_scraped                                  5734 non-null   object 
 2   name                                          5734 non-null   object 
 3   summary                                       5563 non-null   object 
 4   space                                         4571 non-null   object 
 5   description                                   5606 non-null   object 
 6   neighborhood_overview                         3884 non-null   object 
 7   notes                                         3191 non-null   object 
 8   transit                                       3618 non-null   object 
 9   access                                        3642 non-null   o

In [193]:
data_rev.describe(include='all')

id last_scraped  \
count   5.734000e+03         5734   
unique           NaN            2   
top              NaN   2020-05-31   
freq             NaN         3778   
mean    2.461825e+07          NaN   
std     1.209688e+07          NaN   
min     4.952000e+03          NaN   
25%     1.489002e+07          NaN   
50%     2.598272e+07          NaN   
75%     3.500704e+07          NaN   
max     4.356033e+07          NaN   

                                                    name  \
count                                               5734   
unique                                              5615   
top     TRIBE ❤️ Deep Cleaned  2BR @ Great Mall Milpitas   
freq                                                  21   
mean                                                 NaN   
std                                                  NaN   
min                                                  NaN   
25%                                                  NaN   
50%                                                  NaN   
75%                                                  NaN   
max                                                  NaN   

                                                  summary  \
count                                                5563   
unique                                               4920   
top     The HackerHome Network New to the Valley? Tran...   
freq                                                   44   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                                    space  \
count                                                4571   
unique                                               3791   
top     At Zeus, we focus exclusively on 30 day+ stays...   
freq                                                  242   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                              description  \
count                                                5606   
unique                                               5224   
top     Dorm-style house in Silicon Valley. No-host, 5...   
freq                                                   20   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                    neighborhood_overview  \
count                                                3884   
unique                                               3066   
top     *Area Highlights* -Located in the Heart of the...   
freq                                                   28   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                        

In [194]:
counts = reviews['listing_id'].value_counts()


In [195]:
counts.head()

7476637     489
52786       478
10814836    445
19641513    429
13828514    427
Name: listing_id, dtype: int64

## Fusion with neighborhood data

In [196]:
neighbors = pd.read_csv('./data/neighbourhoods.csv')
print(neighbors.shape)
neighbors.head()

(16, 2)


neighbourhood_group    neighbourhood
0                  NaN         Campbell
1                  NaN        Cupertino
2                  NaN           Gilroy
3                  NaN        Los Altos
4                  NaN  Los Altos Hills

In [197]:
data_neighbor = data['neighbourhood'].unique().tolist()
neighbors_hood = neighbors['neighbourhood'].unique().tolist()

In [198]:
same_locations = []

for neighbor in neighbors_hood:
    if neighbor in data_neighbor:
        same_locations.append(neighbor)

In [199]:
same_locations

['Campbell',
 'Cupertino',
 'Los Altos',
 'Los Altos Hills',
 'Mountain View',
 'Palo Alto',
 'Santa Clara',
 'Sunnyvale']

In [200]:
data_neighbor

['Palo Alto',
 'Santa Clara',
 'Mountain View',
 'South San Jose',
 'Cupertino',
 'Sunnyvale',
 'Downtown',
 'Campbell',
 'West Valley',
 'Edenvale',
 nan,
 'Willow Glen',
 'Central San Jose',
 'Los Altos',
 'Berryessa',
 'Cory',
 'College Park',
 'Alum Rock',
 'Cambrian/Pioneer',
 'North San Jose',
 'Los Altos Hills',
 'Burbank/Del Monte',
 'Evergreen',
 'Newhall/Sherwood',
 'Shasta/Hanchett Park',
 'Rose Garden',
 'Five Wounds/Brookwood Terrace',
 'Naglee Park',
 'Forest/Pruneridge',
 'Alviso',
 'Japantown',
 'Stanford',
 'Delmas Park',
 'Chapman/Morse',
 'Menlo Park',
 'Autumn/Montgomery',
 'Vermont/McKendrie']

In [201]:
neighbors_hood

['Campbell',
 'Cupertino',
 'Gilroy',
 'Los Altos',
 'Los Altos Hills',
 'Los Gatos',
 'Milpitas',
 'Monte Sereno',
 'Morgan Hill',
 'Mountain View',
 'Palo Alto',
 'San Jose',
 'Santa Clara',
 'Saratoga',
 'Sunnyvale',
 'Unincorporated Areas']


## ideas list

### methdology for this large scale data?


### which cols we need

- e.g.

  id
  name
  host_id
  host_name

  neighbourhood_group
  neighbourhood

  latitude
  longitude

  room_type

  price

  minimum_nights
  number_of_reviews

  reviews_per_month
  calculated_host_listings_count
  availability_365

### What to compare

- Room Type vs Price
- Price/log price Distribution Plot


### pipleline?



## Codes for refering

- [amenities refer](https://www.kaggle.com/brittabettendorf/predicting-prices-xgboost-feature-engineering)
- [word cloud /weighted words /sentimental analsys / topic modeling](https://www.kaggle.com/brittabettendorf/nlp-on-airbnb-data)


### nextweek

modeling for price prediction


